pkg/reconciler/pipelinerun/pipelinerun_test.go

```diff
@@ -7952,6 +7986,362 @@ spec:
 		})
 	}
 }
+func TestReconciler_PipelineTaskMatrixWithArrayReferences(t *testing.T) {
+	names.TestingSeed()
+
+	task := parse.MustParseV1beta1Task(t, `
+metadata:
+  name: mytask
+  namespace: foo
+spec:
+  params:
+    - name: platform
+    - name: browser
+  steps:
+    - name: echo
+      image: alpine
+      script: |
+        echo "$(params.platform) and $(params.browser)"
+`)
+
+	expectedTaskRuns := []*v1beta1.TaskRun{
```

Comment:
I wonder if we could use for loop to create the `expectedTaskRuns` to reduce the size of the test code? The difference between each case seem to be the `params`, we could use %s in the `mustParseTaskRunWithObjectMeta` to pass the params? 
e.g. https://github.com/tektoncd/pipeline/blob/50fa5f12d4a1958eab6f191d5baf2c434b3ea4f3/pkg/reconciler/pipelinerun/pipelinerun_test.go#L3487

Reply:
We could ignore this comment in this PR, I just feel like it would be helpful if we could clean up it. 
The suggestion in my mind is something like:
```
 data := []struct {
        name     string
        browser  string
        platform string
    }{{
        name:     "pr-platforms-and-browsers-0",
        browser:  "chrome",
        platform: "linux",
    }, {
        name:     "pr-platforms-and-browsers-1",
        browser:  "chrome",
        platform: "mac",
    }}
    expectedTaskRuns := []*v1beta1.TaskRun{}
    for _, d := range data {
        expectedTaskRuns1 = append(expectedTaskRuns,

            mustParseTaskRunWithObjectMeta(t,
                taskRunObjectMeta(d.name, "foo",
                    "pr", "p-dag", "platforms-and-browsers", false),
                fmt.Sprintf(`
            spec:
            params:
            - name: browser
                value: %s
            - name: platform
                value: %s
            serviceAccountName: test-sa
            taskRef:
                name: mytask
                kind: Task
            labels:
                tekton.dev/memberOf: tasks
                tekton.dev/pipeline: p-dag
            `, d.browser, d.platform)),
        )
    }
```

Reply:
Thanks for the suggestion @Yongxuanzhang. I updated this here and think in the future we could refactor all of the tests to use this and perhaps make a  `createExpectedTaskRuns()` function that can abstract some of this even more from a mapping of parameter names and values.  

I also think there's opportunity to create more variables (pr name, namespace, pipelineTaskName, etc  that can be passed between the specs to avoid some of the manual errors and pain points around formatting the yaml syntax. 

cc: @jerop @lbernick